src/proxy-worker.ts

```diff
@@ -0,0 +1,285 @@
+/*
+	Copyright (C) 2017 Balena Ltd.
+
+	This program is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published
+	by the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import { metrics } from '@balena/node-metrics-gatherer';
+import * as Bluebird from 'bluebird';
+import * as dns from 'dns';
+import * as _ from 'lodash';
+import * as net from 'net';
+import * as nodeTunnel from 'node-tunnel';
+import * as winston from 'winston';
+
+import { captureException, device, errors, getLogger, Metrics } from './utils';
+
+const VPN_SERVICE_API_KEY = Buffer.from(process.env.VPN_SERVICE_API_KEY!);
+const VPN_CONNECT_PROXY_PORT = parseInt(
+	process.env.VPN_CONNECT_PROXY_PORT!,
+	10,
+);
+
+const lookupAsync = Bluebird.promisify(dns.lookup);
+
+class Tunnel extends nodeTunnel.Tunnel {
+	private readonly logger: winston.Logger;
+
+	constructor(
+		private readonly instanceId: number,
+		private readonly serviceId: string,
+	) {
+		super();
+		this.logger = getLogger('proxy', this.instanceId, this.serviceId);
+
+		this.on('error', err => {
+			// errors thrown in `this.connect` will appear here
+			if (!(err instanceof errors.HandledTunnelingError)) {
+				this.logger.crit(
+					`failed to connect to device (${err.message || err})\n${err.stack}`,
+				);
+				captureException(err, 'proxy-connect-error');
+			}
+		});
+
+		this.use(this.tunnelToDevice);
+	}
+
+	public connect = (
+		port: number,
+		host: string,
+		client: net.Socket,
+		req: nodeTunnel.Request,
+	) =>
+		Bluebird.try(() => this.parseRequest(req)).then(({ uuid, auth }) =>
+			lookupAsync(`${uuid}.vpn`)
+				.return(true)
+				.catchReturn(false)
+				.then(exists => {
+					if (exists) {
+						// The lookup succeeded so we try to connect
+						this.logger.info(`connecting to ${host}:${port}`);
+						return super.connect(port, host, client, req).tap(socket => {
+							metrics.inc(Metrics.ActiveTunnels);
+							metrics.inc(Metrics.TotalTunnels);
+							socket.on('close', () => {
+								metrics.dec(Metrics.ActiveTunnels);
+							});
+						});
+					} else {
+						// The lookup failed so we try to forward to the correct vpn instance instead
+						return device
+							.getDeviceVpnHost(uuid, auth)
+							.catch(errors.APIError, err => {
```

Comment:
Again depending on how often this is called then this sugar imposes a definite cost

Reply:
like this?

```ts
.catch(err => {
	if (err instanceof errors.APIError) {
		this.logger.crit(
			`error connecting to device ${uuid} on port ${port} (${err.message})`,
		);
		throw new errors.HandledTunnelingError(err.message);
	}
	throw err;
})
```